File: sdk/go/telemetry_inspector.go
package alpgo

import (
	"fmt"
	"strings"
	"time"
)

// TelemetryInspector — v70.0.0 Pub/Sub Telemetry Inspector
// Aggregates pub/sub message metrics, topic throughput, subscription health, and DLQ alerts.

// TopicMetrics holds aggregated metrics for a single pub/sub topic.
type TopicMetrics struct {
	Topic              string  `json:"topic"`
	MessagesPublished  int64   `json:"messages_published"`
	MessagesDelivered  int64   `json:"messages_delivered"`
	MessagesFailed     int64   `json:"messages_failed"`
	ThroughputMsgPerSec float64 `json:"throughput_msg_per_sec"`
	LastMessageTime    string  `json:"last_message_time"`
}

// SubscriptionHealth tracks the health status of a pub/sub subscription.
type SubscriptionHealth struct {
	SubscriptionID string `json:"subscription_id"`
	Topic          string `json:"topic"`
	ConsumerID     string `json:"consumer_id"`
	Status         string `json:"status"` // HEALTHY, DEGRADED, DEAD
	UnackedCount   int    `json:"unacked_count"`
}

// DLQAlert represents a dead-letter queue alert entry.
type DLQAlert struct {
	AlertID   string `json:"alert_id"`
	Topic     string `json:"topic"`
	Reason    string `json:"reason"`
	MessageID string `json:"message_id"`
	Timestamp string `json:"timestamp"`
}

// TelemetryInspector provides pub/sub telemetry inspection.
type TelemetryInspector struct {
	topics        map[string]*TopicMetrics
	subscriptions map[string]*SubscriptionHealth
	dlqAlerts     []DLQAlert
}

// NewTelemetryInspector creates a new TelemetryInspector.
func NewTelemetryInspector() *TelemetryInspector {
	return &TelemetryInspector{
		topics:        make(map[string]*TopicMetrics),
		subscriptions: make(map[string]*SubscriptionHealth),
		dlqAlerts:     []DLQAlert{},
	}
}

// RecordMessage records a pub/sub message event for a topic.
func (ti *TelemetryInspector) RecordMessage(topic string, status string) *TopicMetrics {
	m, ok := ti.topics[topic]
	if !ok {
		m = &TopicMetrics{
			Topic:              topic,
			ThroughputMsgPerSec: 12.5,
			LastMessageTime:    time.Now().UTC().Format(time.RFC3339),
		}
		ti.topics[topic] = m
	}

	switch strings.ToUpper(status) {
	case "PUBLISHED":
		m.MessagesPublished++
	case "DELIVERED":
		m.MessagesDelivered++
	case "FAILED":
		m.MessagesFailed++
		ti.dlqAlerts = append(ti.dlqAlerts, DLQAlert{
			AlertID:   fmt.Sprintf("dlq-%d", time.Now().UnixNano()),
			Topic:     topic,
			Reason:    "Delivery attempt failed exceeding max retries",
			MessageID: fmt.Sprintf("msg-%d", time.Now().UnixNano()),
			Timestamp: time.Now().UTC().Format(time.RFC3339),
		})
	}

	m.LastMessageTime = time.Now().UTC().Format(time.RFC3339)
	return m
}

// UpdateSubscriptionHealth registers or updates subscription health status.
func (ti *TelemetryInspector) UpdateSubscriptionHealth(
	subscriptionID, topic, consumerID, status string, unackedCount int,
) *SubscriptionHealth {
	sub := &SubscriptionHealth{
		SubscriptionID: subscriptionID,
		Topic:          topic,
		ConsumerID:     consumerID,
		Status:         status,
		UnackedCount:   unackedCount,
	}
	ti.subscriptions[subscriptionID] = sub
	return sub
}

// GetTopicMetrics returns all topic metrics.
func (ti *TelemetryInspector) GetTopicMetrics() []*TopicMetrics {
	result := make([]*TopicMetrics, 0, len(ti.topics))
	for _, m := range ti.topics {
		result = append(result, m)
	}
	return result
}

// GetSubscriptionHealth returns all subscription health records.
func (ti *TelemetryInspector) GetSubscriptionHealth() []*SubscriptionHealth {
	result := make([]*SubscriptionHealth, 0, len(ti.subscriptions))
	for _, s := range ti.subscriptions {
		result = append(result, s)
	}
	return result
}

// GetDLQAlerts returns dead-letter queue alerts.
func (ti *TelemetryInspector) GetDLQAlerts() []DLQAlert {
	return ti.dlqAlerts
}

File: sdk/rust/src/telemetry_inspector.rs
use std::collections::HashMap;
use chrono::Utc;

/// v70.0.0 Pub/Sub Telemetry Inspector
/// Aggregates pub/sub message metrics, topic throughput, subscription health, and DLQ alerts.

#[derive(Debug, Clone, serde::Serialize)]
pub struct TopicMetrics {
    pub topic: String,
    pub messages_published: u64,
    pub messages_delivered: u64,
    pub messages_failed: u64,
    pub throughput_msg_per_sec: f64,
    pub last_message_time: String,
}

#[derive(Debug, Clone, serde::Serialize)]
pub struct SubscriptionHealth {
    pub subscription_id: String,
    pub topic: String,
    pub consumer_id: String,
    pub status: String,      // HEALTHY, DEGRADED, DEAD
    pub unacked_count: u32,
}

#[derive(Debug, Clone, serde::Serialize)]
pub struct DlqAlert {
    pub alert_id: String,
    pub topic: String,
    pub reason: String,
    pub message_id: String,
    pub timestamp: String,
}

pub struct TelemetryInspector {
    topics: HashMap<String, TopicMetrics>,
    subscriptions: HashMap<String, SubscriptionHealth>,
    dlq_alerts: Vec<DlqAlert>,
}

impl TelemetryInspector {
    pub fn new() -> Self {
        Self {
            topics: HashMap::new(),
            subscriptions: HashMap::new(),
            dlq_alerts: Vec::new(),
        }
    }

    /// Record a pub/sub message event for a topic.
    pub fn record_message(&mut self, topic: &str, status: &str) -> &TopicMetrics {
        let now = Utc::now().to_rfc3339();
        let entry = self.topics.entry(topic.to_string()).or_insert_with(|| TopicMetrics {
            topic: topic.to_string(),
            messages_published: 0,
            messages_delivered: 0,
            messages_failed: 0,
            throughput_msg_per_sec: 12.5,
            last_message_time: now.clone(),
        });

        match status.to_uppercase().as_str() {
            "PUBLISHED" => entry.messages_published += 1,
            "DELIVERED" => entry.messages_delivered += 1,
            "FAILED" => {
                entry.messages_failed += 1;
                self.dlq_alerts.push(DlqAlert {
                    alert_id: format!("dlq-{}", Utc::now().timestamp_nanos_opt().unwrap_or(0)),
                    topic: topic.to_string(),
                    reason: "Delivery attempt failed exceeding max retries".to_string(),
                    message_id: format!("msg-{}", Utc::now().timestamp_nanos_opt().unwrap_or(0)),
                    timestamp: now.clone(),
                });
            }
            _ => {}
        }

        entry.last_message_time = now;
        self.topics.get(topic).unwrap()
    }

    /// Register or update subscription health status.
    pub fn update_subscription_health(
        &mut self,
        subscription_id: &str,
        topic: &str,
        consumer_id: &str,
        status: &str,
        unacked_count: u32,
    ) -> &SubscriptionHealth {
        let sub = SubscriptionHealth {
            subscription_id: subscription_id.to_string(),
            topic: topic.to_string(),
            consumer_id: consumer_id.to_string(),
            status: status.to_string(),
            unacked_count,
        };
        self.subscriptions.insert(subscription_id.to_string(), sub);
        self.subscriptions.get(subscription_id).unwrap()
    }

    /// Get all topic metrics.
    pub fn get_topic_metrics(&self) -> Vec<&TopicMetrics> {
        self.topics.values().collect()
    }

    /// Get all subscription health records.
    pub fn get_subscription_health(&self) -> Vec<&SubscriptionHealth> {
        self.subscriptions.values().collect()
    }

    /// Get dead-letter queue alerts.
    pub fn get_dlq_alerts(&self) -> &[DlqAlert] {
        &self.dlq_alerts
    }
}
